topology/probes/blockdev/blockdev.go

```diff
@@ -0,0 +1,515 @@
+/*
+ * Copyright (C) 2019 Red Hat, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy ofthe License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specificlanguage governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package blockdev
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"os"
+	"os/exec"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+	"unicode/utf8"
+
+	"github.com/skydive-project/skydive/common"
+	"github.com/skydive-project/skydive/graffiti/graph"
+	"github.com/skydive-project/skydive/probe"
+	"github.com/skydive-project/skydive/topology"
+	"github.com/skydive-project/skydive/topology/probes"
+	tp "github.com/skydive-project/skydive/topology/probes"
+)
+
+const blockdevGroupName string = "block devices"
+
+// Types returned from lsblk JSON
+const multipathType string = "mpath"
+const lvmType string = "lvm"
+const blockGroupType string = "cluster"
+
+// Types for skydive nodes
+const lvmNodeType string = "blockdevlvm"
+const blockdevNodeType string = "blockdev"
+const leafNodeType string = "blockdevleaf"
+
+const managerType string = "blockdev"
+
+// BlockDevice used for JSON parsing
+// Prior to version 2.33 lsblk generated JSON that encodes all values
+// as strings.  Version 2.33 changed how integers and booleans are
+// encoded.  Using json.RawMessage allows the code deal with either
+// encoding.
+type BlockDevice struct {
+	Children []BlockDevice `json:"children"`
+
+	Alignment    json.RawMessage `json:"alignment"` //int64
```

Comment:
An alternative would be :
- use json.Unmarshal to a map[string]interface{}
- then use https://godoc.org/github.com/mitchellh/mapstructure#WeakDecode to map to a `BlockDevice` struct. It converts strings to ints (and more) so it could handle a change of type of a member.

It's really up to you :-) We can merge it without it

Reply:
I think I've implemented it the way you've suggested.  It looks better to me - please have a look and let me know if I got it right. :-)  Will push soon.

Reply:
Ok - I see your 2nd commit looks better than what I came up with... I'll cherry-pick and test.

Reply:
@lebauce I squashed your JSON parser changes (hope that is ok) and cherry-picked the "lsblk" path changes.

where are you thinking we should put the fake lsblk?